### ISSUE TITLE
toc not working for github_document

### ISSUE THREAD

**Issue by jennybc** [rstudio/rmarkdown#1311]
Table of contents is currently not working for the `github_document` output format.

I'm using the yaml frontmatter that is given here:

https://rmarkdown.rstudio.com/github_document_format.html#table_of_contents

When I render this, I get no table of contents. If I switch to `md_document`, I do.

````
---
title: "Habits"
output:
  github_document:
    toc: true
    toc_depth: 2
---

## Section 1

This is an R Markdown document.

```{r cars}
summary(cars)
```

## Section 2

More stuff

```{r}
1:3 ^ 2
```
````

**Comment by rich-iannone** [comment]
@jennybc I'm looking into this now. Seems to be a Pandoc 2.0+ issue, occurring in the conversion of the intermediate `*.utf8.md` doc with the Pandoc arg `--to gfm-ascii_identifiers`.

**Comment by rich-iannone** [comment]
It seems that Pandoc's `--to gfm` (which replaces `--to markdown_github`) hasn't implemented toc creation (with `--toc` or `--table-of-contents`). 

This OP for the (closed) issue https://github.com/jgm/pandoc/issues/4132 indicated that toc creation didn't work. 

There is a similar issue in this repo (https://github.com/rstudio/rmarkdown/issues/1211), which is more relevant now that there is a move to Pandoc 2.0+.

With the file `toc_testing_github_document.utf8.md`, I did some basic command line testing with pandoc to verify whether a toc is generated with `--to gfm`. Here is that input file:

````
---
title: "Habits"
output:
  github_document:
    toc: true
    toc_depth: 2
---

## Section 1

This is an R Markdown document.


```r
summary(cars)
```

```
##      speed           dist       
##  Min.   : 4.0   Min.   :  2.00  
##  1st Qu.:12.0   1st Qu.: 26.00  
##  Median :15.0   Median : 36.00  
##  Mean   :15.4   Mean   : 42.98  
##  3rd Qu.:19.0   3rd Qu.: 56.00  
##  Max.   :25.0   Max.   :120.00
```

## Section 2

More stuff


```r
1:3 ^ 2
```

```
## [1] 1 2 3 4 5 6 7 8 9
```
````

I tried various combinations of the following pandoc call in terminal, but couldn't get a toc to appear like when `--to markdown_github` is used.

```
/Applications/RStudio.app/Contents/MacOS/pandoc/pandoc +RTS -K512m -RTS toc_testing_github_document.utf8.md --to markdown_github --from markdown+autolink_bare_uris+tex_math_single_backslash --output toc_testing_github_document-options-testing.md --standalone --toc --toc-depth 2 --template /Library/Frameworks/R.framework/Versions/3.4/Resources/library/rmarkdown/rmarkdown/templates/github_document/resources/default.md 
```

The Pandoc documentation doesn't explicitly mention that `--toc` doesn't work for `--to gfm`. I think the only recourse is for me to post another issue to `jgm/pandoc`.

**Comment by jennybc** [comment]
Anecdotally, I have noticed that I get better results re: producing markdown for display on GitHub with `md_document` and `-t commonmark` (vs `github_document`) these days. That's what I do in reprex now. I think it's possible that the `github_document` output format should just get a refresh along those lines.

I know that [GitHub moved to a formal spec based on CommonMark](https://blog.github.com/2017-03-14-a-formal-spec-for-github-flavored-markdown/) about a year ago, but I don't know if Pandoc's `-t gfm` kept pace and changed accordingly (?). As a Pandoc user, it sort of feels like "no".

**Comment by jjallaire** [comment]
@rich-iannone Maybe we should change the implementation of github_document to use a variation of md_document as @jennybc suggests?

**Comment by rich-iannone** [comment]
@jjallaire that seems like the best path forward. It’s unknown whether Pandoc’s `-t gfm` will improve in the short term and `t commonmark` should give us everything we need. 

@jennybc thanks for the suggestion, I’ll work on implementing and testing.

**Comment by jennybc** [comment]
While I am sharing anecdote, I also use `-f markdown-implicit_figures` with `-t commonmark`. What little backstory I have on that is [here](https://groups.google.com/forum/#!searchin/pandoc-discuss/jenny%7Csort:date/pandoc-discuss/kl6Nbr-dcxk/LzafpbAOBAAJ). I was having trouble with line breaks in reprex and this helped. I have had various woes with line wrapping (esp. wrt Pandoc >= 2 vs < 2), but I will spare you the details. We can chat via Slack if you want to follow up.

**Comment by gaskova** [comment]
Hello, I need to use a table of contents in my R Markdown document but it doesn't work. Any solution yet? I am sorry, I don't understand if it is fixed, or not, from this conversation. It seems to me that not. I am a beginner. Thank you very much.

**Comment by EmilHvitfeldt** [comment]
I just tried to knit the example document(posted by Jenny). And I get a working toc.

pandoc: 1.19.2.1
rmarkdown: 1.11     
knitr: 1.21  

``` r
rmarkdown::pandoc_version()
#> [1] '1.19.2.1'
sessioninfo::session_info()
#> ─ Session info ──────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 3.5.1 (2018-07-02)
#>  os       macOS High Sierra 10.13.6   
#>  system   x86_64, darwin15.6.0        
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       America/Los_Angeles         
#>  date     2018-12-27                  
#> 
#> ─ Packages ──────────────────────────────────────────────────────────────
#>  package     * version date       lib source        
#>  assertthat    0.2.0   2017-04-11 [1] CRAN (R 3.5.0)
#>  cli           1.0.1   2018-09-25 [1] CRAN (R 3.5.0)
#>  crayon        1.3.4   2017-09-16 [1] CRAN (R 3.5.0)
#>  digest        0.6.18  2018-10-10 [1] CRAN (R 3.5.0)
#>  evaluate      0.12    2018-10-09 [1] CRAN (R 3.5.0)
#>  highr         0.7     2018-06-09 [1] CRAN (R 3.5.0)
#>  htmltools     0.3.6   2017-04-28 [1] CRAN (R 3.5.0)
#>  knitr         1.21    2018-12-10 [1] CRAN (R 3.5.1)
#>  magrittr      1.5     2014-11-22 [1] CRAN (R 3.5.0)
#>  Rcpp          1.0.0   2018-11-07 [1] CRAN (R 3.5.0)
#>  rmarkdown     1.11    2018-12-08 [1] CRAN (R 3.5.0)
#>  sessioninfo   1.1.1   2018-11-05 [1] CRAN (R 3.5.0)
#>  stringi       1.2.4   2018-07-20 [1] CRAN (R 3.5.0)
#>  stringr       1.3.1   2018-05-10 [1] CRAN (R 3.5.0)
#>  withr         2.1.2   2018-03-15 [1] CRAN (R 3.5.0)
#>  xfun          0.4     2018-10-23 [1] CRAN (R 3.5.0)
#>  yaml          2.2.0   2018-07-25 [1] CRAN (R 3.5.0)
#> 
#> [1] /Library/Frameworks/R.framework/Versions/3.5/Resources/library
```

<sup>Created on 2018-12-27 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by cderv** [comment]
@EmilHvitfeldt the issue is with pandoc 2.0 and later as @rich-iannone stated. You have Pandoc 1, so it is working fine. 

if fact, it seems to be an issue in pandoc regarding the `--toc` and `gfm` (https://github.com/jgm/pandoc/issues/5172) and it seems from the comment that `gfm` use indeed commonmark (cc @jennybc ) and issue comes from here. 

Regarding pandoc version behaviour, rmarkdown use `-t gfm` if pandoc version is above 2 
https://github.com/rstudio/rmarkdown/blob/b209cdc7f25e126c758aab5048849a2909f3bb1a/R/github_document.R#L35-L37
`-t markdown_github`, the old format from pandoc 1, is working fine. It could still be use with pandoc 2, and toc is also working in this case. This could be a workraround (to use with `md_document`)

At the end, I think this should be fixed in pandoc and it will be ok in rmarkdown, or rmarkdown should not rely on `gfm` format for `github_document` but a more custom variation of [markdown pandoc format](http://pandoc.org/MANUAL.html#markdown-variants).

**Comment by behrman** [comment]
@rich-iannone and @jennybc, I'm putting together a class that starts on Mon. We have been using github_document extensively with tables of contents. What do you recommend now? Is there a way to obtain tables of contents while this bug is being fixed?

**Comment by cderv** [comment]
The issue seems to be fixed in last pandoc (with https://github.com/jgm/pandoc/commit/c4c3fca50009ef5d8b3b34aaca64007ebe70f0a8). Don't know when it will come ouput and be part of Rstudio.

In the meantime, if you have pandoc < 2, there is no issue. If you have pandoc 2.0+, you could use `md_document()` with `variant = "markdown_github"` which is the old format. Should work.

**Comment by jennybc** [comment]
Ha! I was re-investigating this too @cderv!

Yeah, this works:

```
---
output:
  md_document:
    variant: markdown_github
    toc: true
    toc_depth: 2
---

## Heading 1

blah blah

## Heading 2

blah blah
```

and you just have to ignore this warning:

```
[WARNING] Deprecated: markdown_github. Use gfm instead.
```

**Comment by behrman** [comment]
Many thanks, @cderv and @jennybc. It looks as though the bug in pandoc may have been fixed 2 days ago.

I just tried the fix on a simple document with no chunks using RStudio 1.2.1206. With github_document, it took 2 seconds to compile. With md_document, it to 2 minutes and 24 seconds.

Do you have any idea how long it will take the fixed pandoc to work its way into the preview RStudio?

**Comment by jennybc** [comment]
> I just tried the fix on a simple document with no chunks using RStudio 1.2.1206. With github_document, it took 2 seconds to compile. With md_document, it to 2 minutes and 24 seconds.

That sounds very peculiar. Can you reproduce that behaviour reliably? I have had RStudio occasionally become very sluggish and I wonder if your `md_document` experiment perhaps coincided with a similar phenomenon.

> Do you have any idea how long it will take the fixed pandoc to work its way into the preview RStudio?

I predict it will be "a while" because that will first have to make it into a Pandoc release and then that version of Pandoc would need to make it into RStudio Preview or release. You can't wait on that to figure out the `.Rmd` YAML you'll recommend in your course starting soon.

**Comment by behrman** [comment]
I want to thank you for recommending github_document a couple of years ago when it was new. This greatly improved the workflow of students submitting assignments and the staff reading them.

If it will be "a while" for the new Pandoc to work its way into RStudio, perhaps we should revert to an earlier version of Pandoc without this problem. Any pointers in how to go about doing that?

**Comment by behrman** [comment]
I have a workaround to use the old Pandoc. Creating an .Rprofile with with the following line

    Sys.setenv(RSTUDIO_PANDOC="path to old pandoc") 

produces tables of contents without changing code. Please let me know if there's a better way to use an old Pandoc.

**Comment by duffau** [comment]
I'v had the same problem as with `toc: true` not wotking for `output: github_document`. 

Installing [Pandoc 2.6](https://pandoc.org/installing.html) fixed the issue for me.

The release note [release notes of Pandoc 2,6](https://pandoc.org/releases.html) state that the the issue with the `--toc` flag in  `gfm`-mode is fixed: 
> Implement --toc (writerTableOfContents) in commonmark/gfm writers (#5172).

**Comment by mstuart1** [comment]
Updating [pandoc](https://pandoc.org/installing.html) solved this issue for me.

**Comment by cderv** [comment]
Just passing by here and I think the original issue is now fixed by pandoc since pandoc 2.6. Pandoc 2.7.3 is included with RStudio now.

BTW, There is also an improved `gfm` format  (and new commonmark_x). From pandoc 2.10.1, announcement: 
> For rendering, we rely on the markdown writer, which when supplied with appropriate extensions and options generates correct commonmark.

This should improve the rendering of `gfm` document. But this is another topic.

Anyway, I think we could close this issue now. What are you thoughts on the original issue @jennybc ?

**Comment by jennybc** [comment]
I'm happy for you to close this if the original problem (lack of TOC) has self-resolved.

I am interested in getting to the bottom of how best to produce markdown for display on GitHub. But that's really a separate issue and maybe just be a matter of checking / documenting things? But we definitely went through a period of years where `github_document` was NOT, in fact, the best way to produce GitHub-flavoured markdown.

**Comment by behrman** [comment]
@cderv and @jennybc, I've just been having students separately install a newer version of Pandoc without the TOC problem. If Pandoc 2.7.3 is now included with RStudio, I believe that should resolve the problem.

Thanks

**Comment by cderv** [comment]
> I am interested in getting to the bottom of how best to produce markdown for display on GitHub. But that's really a separate issue and maybe just be a matter of checking / documenting things?

I am interested too in this, all the more with the new pandoc 2.10.1 version that has improve commonmark support. If pandoc gfm format is the correct one to produce GitHub flavored markdown, it should be pretty direct. I’ll maybe open a new issue to track this and look into this.
